File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/nowplaying/storage/NowPlayingRepository.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.nowplaying.storage;

import com.lasthopesoftware.bluewater.client.library.access.ILibraryStorage;
import com.lasthopesoftware.bluewater.client.library.access.ISpecificLibraryProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.access.stringlist.FileStringListUtilities;
import com.lasthopesoftware.messenger.promises.Promise;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by david on 1/29/17.
 */
public class NowPlayingRepository implements INowPlayingRepository {

	private static final Map<Integer, NowPlaying> nowPlayingCache = new ConcurrentHashMap<>();

	private final ISpecificLibraryProvider libraryProvider;
	private final ILibraryStorage libraryRepository;

	private volatile int libraryId = -1;

	public NowPlayingRepository(ISpecificLibraryProvider libraryProvider, ILibraryStorage libraryRepository) {
		this.libraryProvider = libraryProvider;
		this.libraryRepository = libraryRepository;
	}

	@Override
	public Promise<NowPlaying> getNowPlaying() {
		if (nowPlayingCache.containsKey(libraryId))
			return new Promise<>(nowPlayingCache.get(libraryId));

		return
			libraryProvider
				.getLibrary()
				.then(library -> {
					libraryId = library.getId();

					final String savedTracksString = library.getSavedTracksString();
					if (savedTracksString == null || savedTracksString.isEmpty()) {
						final NowPlaying nowPlaying =
							new NowPlaying(
								library.getNowPlayingId(),
								library.getNowPlayingProgress(),
								library.isRepeating());

						nowPlayingCache.put(libraryId, nowPlaying);

						return new Promise<>(nowPlaying);
					}

					return
						FileStringListUtilities
							.promiseParsedFileStringList(savedTracksString)
							.next(files -> {
								final NowPlaying nowPlaying =
									new NowPlaying(
										files,
										library.getNowPlayingId(),
										library.getNowPlayingProgress(),
										library.isRepeating());

								nowPlayingCache.put(libraryId, nowPlaying);

								return nowPlaying;
							});
				});
	}

	@Override
	public Promise<NowPlaying> updateNowPlaying(NowPlaying nowPlaying) {
		if (libraryId < 0)
			return getNowPlaying().then(np -> updateNowPlaying(nowPlaying));

		nowPlayingCache.put(libraryId, nowPlaying);

		libraryProvider
			.getLibrary()
			.then(library -> {
				library.setNowPlayingId(nowPlaying.playlistPosition);
				library.setNowPlayingProgress(nowPlaying.filePosition);
				library.setRepeating(nowPlaying.isRepeating);

				return
					FileStringListUtilities
						.promiseSerializedFileStringList(nowPlaying.playlist)
						.then(serializedPlaylist -> {
							library.setSavedTracksString(serializedPlaylist);

							return libraryRepository.saveLibrary(library);
						});
			});

		return new Promise<>(nowPlaying);
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/nowplaying/storage/NowPlayingRepository.kt
package com.lasthopesoftware.bluewater.client.playback.nowplaying.storage

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.access.stringlist.FileStringListUtilities.promiseParsedFileStringList
import com.lasthopesoftware.bluewater.client.browsing.files.access.stringlist.FileStringListUtilities.promiseSerializedFileStringList
import com.lasthopesoftware.bluewater.client.browsing.library.access.ManageLibraries
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.ProvideSelectedLibraryId
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryNowPlayingValues
import com.lasthopesoftware.policies.caching.LruPromiseCache
import com.lasthopesoftware.promises.extensions.keepPromise
import com.lasthopesoftware.promises.extensions.toPromise
import com.namehillsoftware.handoff.promises.Promise

class NowPlayingRepository(
	private val selectedLibraryId: ProvideSelectedLibraryId,
	private val libraryManager: ManageLibraries
) : ManageNowPlayingState {

	private val serializedTracksCache by lazy { LruPromiseCache<List<ServiceFile>, String>(1) }

	override fun promiseActiveNowPlaying(): Promise<NowPlaying?> =
		selectedLibraryId
			.promiseSelectedLibraryId()
			.eventually { it?.let(::promiseNowPlaying).keepPromise() }

	override fun promiseNowPlaying(libraryId: LibraryId): Promise<NowPlaying?> =
		libraryManager
			.promiseNowPlayingValues(libraryId)
			.eventually { library ->
				library?.run {
					savedTracksString
						?.takeIf { it.isNotEmpty() }
						?.let(::promiseParsedFileStringList)
						?.then { files ->
							val nowPlaying = NowPlaying(
								libraryId,
								if (files is List<*>) files as List<ServiceFile> else ArrayList(files),
								nowPlayingId,
								nowPlayingProgress,
								isRepeating
							)
							nowPlaying
						}
						?: NowPlaying(
							libraryId,
							emptyList(),
							nowPlayingId,
							nowPlayingProgress,
							isRepeating
						).toPromise()
				}.keepPromise()
			}

	override fun updateNowPlaying(nowPlaying: NowPlaying): Promise<NowPlaying> {
		val libraryId = nowPlaying.libraryId

		return Promise.Proxy { cp ->
			serializedTracksCache
				.getOrAdd(nowPlaying.playlist, ::promiseSerializedFileStringList)
				.also(cp::doCancel)
				.eventually { serializedPlaylist ->
					nowPlaying.run {
						libraryManager
							.updateNowPlaying(
								LibraryNowPlayingValues(
									id = libraryId.id,
									nowPlayingId = playlistPosition,
									nowPlayingProgress = filePosition,
									savedTracksString = serializedPlaylist,
									isRepeating = isRepeating
								)
							)
							.also(cp::doCancel)
					}
				}
				.then { _ -> nowPlaying }
		}
	}
}
